csrc/lower_thread_predicate.h

```diff
@@ -52,6 +52,22 @@ class TORCH_CUDA_CU_API ThreadPredicateMap {
     ParallelTypeBitmap limited_types;
     // Parallel types where only one thread/block is enough.
     ParallelTypeBitmap redundant_types;
+
+    // when a leaf domain is merged from concretized broadcast root domain, only
+    // part of thread/block do the write to gmem is enough e.g. [B1,I2,B3] is
+    // merged to [B1*I2*B3] and parallelized by blockIdx.x. The write pattern
+    // should be: write every len(B3) blocks of the first len(I2) * len(B3)
+    // blocks. write_stride_less= ( ( 1 * T1.size[2] ) * T0.size[0] ).
+    // write_stride_mod= ( 1 * T1.size[2] ).
+    // generated condition is: blockIdx.x < write_stride_less && blockIdx.x %
+    // write_stride_mod < 1
+    // Another example, [I1, B2, I3] merged to [I1*B2*I3], the condition is:
+    // blockIdx.x % (len(B2)*len(I3)) < len(I3).
```

Comment:
I think this comment is indicating that we're thinking of cases where we have a merge with the outer being broadcast and inner not being broadcast. The point being if we have this broadcast axis getting peeled off (a broadcast axis that's resolved inlined) then we only need to write for the first pass of the inner side of that merge.

I think this only works if the iter domain doesn't have "zero merged in", so as long as this is currently limited to global memory instances then it should be valid but wouldn't necessarily be valid, for example, if it was done on shared memory.

Reply:
The generated predication does not depend on the order of the merge, it only depends on the order of the root domains, e.g. [broadcast, iter] or [iter, broadcast]. Current fix is only for global memory. what's the meaning of "zero merged in"?

Reply:
zero merged in is currently only applicable to smem and local, but for multi-gpu it will get extended. It's indicative of the removal of allocation associated to locality across parallelization. i.e. if something is local memory it will remove the portion of the allocation bound to block and grid dims.